ch32fun/ch32fun.h

```diff
@@ -817,11 +830,41 @@ extern "C" {
 // and take two cycles, so you typically would use 0, 2, 4, etc.
 #define ADD_N_NOPS( n ) asm volatile( ".rept " #n "\nc.nop\n.endr" );
 
+#define FUN_HIGH 0x1
+#define FUN_LOW 0x0
+#if defined(CH59x)
+#define GPIOA_ResetBits(pin)      (R32_PA_CLR |= pin)
+#define GPIOA_SetBits(pin)        (R32_PA_OUT |= pin)
+#define GPIOA_InverseBits(pin)    (R32_PA_OUT ^= pin)
+#define GPIOA_ReadPortPin(pin)    (R32_PA_PIN & (pin))
+#define GPIOB_ResetBits(pin)      (R32_PB_CLR |= pin)
+#define GPIOB_SetBits(pin)        (R32_PB_OUT |= pin)
+#define GPIOB_InverseBits(pin)    (R32_PB_OUT ^= pin)
+#define GPIOB_ReadPortPin(pin)    (R32_PB_PIN & (pin))
+#define GPIO_ResetBits(pin)       { if(pin & PB) GPIOB_ResetBits(pin); else GPIOA_ResetBits(pin); }
+#define GPIO_SetBits(pin)         { if(pin & PB) GPIOB_SetBits(pin); else GPIOA_SetBits(pin); }
+#define GPIO_InverseBits(pin)     { if(pin & PB) GPIOB_InverseBits(pin); else GPIOA_InverseBits(pin); }
+#define funDigitalRead(pin)       ( (pin & PB) ? GPIOB_ReadPortPin(pin) : GPIOA_ReadPortPin(pin) )
+#define funDigitalWrite( pin, value ) { if(value==FUN_HIGH){GPIO_SetBits(pin);} else if(value==FUN_LOW){GPIO_ResetBits(pin);} }
+
+#define GPIO_ModeCfg(pd_drv, pu, dir, pin, mode) { switch(mode) { \
+																										case GPIO_ModeIN_Floating: \
+																											pd_drv &= ~pin; pu &= ~pin; dir &= ~pin; break; \
+																										case GPIO_ModeIN_PU: \
+																											pd_drv &= ~pin; pu |= pin; dir &= ~pin; break; \
+																										case GPIO_ModeIN_PD: \
+																											pd_drv |= pin; pu &= ~pin; dir &= ~pin; break; \
+																										case GPIO_ModeOut_PP_5mA: \
+																											pd_drv &= ~pin; dir |= pin; break; \
+																										case GPIO_ModeOut_PP_20mA: \
+																											pd_drv |= pin; dir |= pin; break; \
+																									} }
+#define funPinMode( pin, mode ) { if(pin & PB) GPIO_ModeCfg(R32_PB_PD_DRV, R32_PB_PU, R32_PB_DIR, pin, mode) \
```

Comment:
This looks like it could be simplified where you would have something like R32_PA_PD_DRV+((pin&0xF0)>>1)

Logic: 
PA = 0x400010B4
PB = 0x400010D4
Difference = 0x20.
Type: uint32_t
so they are 8 units apart
The port number is stored in 0xF0, so 16 units apart.

You can check my logic.

Reply:
I more or less did exactly this, I don't see any simplifications anymore (f475b81)